executions/graphql-kotlin-transaction-batcher/src/main/kotlin/com/expediagroup/graphql/transactionbatcher/transaction/TransactionBatcher.kt

```diff
@@ -0,0 +1,97 @@
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.reflect.KClass
+
+/**
+ * Holds logic to apply batching, deduplication and caching of [BatchableTransaction]
+ * if no [TransactionBatcherCache] implementation is provided it will use [DefaultTransactionBatcherCache]
+ */
+class TransactionBatcher(
+    private val cache: TransactionBatcherCache = DefaultTransactionBatcherCache()
+) {
+
+    val batch = ConcurrentHashMap<
+        KClass<out TriggeredPublisher<Any, Any>>,
+        BatchEntryValue
+        >()
+
+    /**
+     * adds a transaction [input] to the batch along with the [triggeredPublisher] instance that will receive the [BatchableTransaction]
+     * deduplication will be based on [transactionKey] which by default is the toString() representation of [input]
+     * batching will be based on the implementation of [TriggeredPublisher]
+     * this method returns a reference to a [CompletableFuture] which is a field of the [BatchableTransaction] that was just
+     * added into the queue
+     */
+    @Suppress("UNCHECKED_CAST")
+    fun <TInput : Any, TOutput : Any> batch(
+        input: TInput,
+        transactionKey: String = input.toString(),
+        triggeredPublisher: TriggeredPublisher<TInput, TOutput>
+    ): CompletableFuture<TOutput> {
+        val publisherClass = (triggeredPublisher as TriggeredPublisher<Any, Any>)::class
+        var future = CompletableFuture<TOutput>()
+        batch.computeIfPresent(publisherClass) { _, batchEntry ->
+            batchEntry
+                .transactions[transactionKey]
+                ?.let { matchedTransaction ->
+                    future = matchedTransaction.future as CompletableFuture<TOutput>
+                }
+                ?: run {
+                    batchEntry.transactions[transactionKey] = BatchableTransaction(
+                        input,
+                        future as CompletableFuture<Any>,
+                        transactionKey
+                    )
+                }
+            batchEntry
+        }
+        batch.computeIfAbsent(publisherClass) {
+            BatchEntryValue(
+                triggeredPublisher,
+                linkedMapOf(
+                    transactionKey to BatchableTransaction(
+                        input,
+                        future as CompletableFuture<Any>,
+                        transactionKey
+                    )
+                )
+            )
+        }
+        return future
+    }
+
+    /**
+     * Trigger concurrently and asynchronously the instances of [TriggeredPublisher] that the [batch] holds
+     * at the end clear the queue
+     */
+    @Synchronized fun dispatch() {
+        batch.values.forEach { (triggeredPublisher, transactions) ->
+            triggeredPublisher.trigger(
+                transactions.map(Map.Entry<String, BatchableTransaction<Any, Any>>::value),
```

Comment:
why not just `transactions.values` (collection) or `transaction.values.toList()`?